### ISSUE TITLE
About  cosine similarity.

### ISSUE THREAD

**Issue by 893018991** [ejkim47/bridging-gap-wsol#2]
Thank you for sharing this awesome code.

I have a little question about the  cosine similarity.
![图片](https://user-images.githubusercontent.com/35445113/232770466-bcdb07e0-39b3-4087-811a-35ca4a73c5f2.png)

As shown, I saw in the paper about cosine similarity, but in the code I see: "sim = F.conv2d(input=x_normalized, weight=weight_normalized.view(*weight_normalized.shape, 1, 1))".
Can you explain it to me?
Thank you.

**Comment by ejkim47** [comment]
Thank you for looking into my code!

Before calculating the cosine similarity, both the feature map (x) and weight are normalized in the following code.

> x_normalized = F.normalize(unerased_x, dim=1)
        weight_normalized = F.normalize(self.fc.weight, dim=1)
        sim = F.conv2d(input=x_normalized, weight=weight_normalized.view(*weight_normalized.shape, 1, 1))

The above code calculates the cosine similarity between x and the weight matrix. 
This implementation allows to parallelize the computation of cosine similarity.

Thank you.

**Comment by 893018991** [comment]
Thank you for answering my question, I understand now.
